### PR TITLE
[Card Games]: Imperative mood for docstring start

### PR DESCRIPTION
**Pull request by Metallifax**
As per title. Addresses #2958

### CODE REVIEW

**Comment by Metallifax** [comment]
@BethanyG As censored for T.V Forrest Gump once said: it happens. I went ahead and installed pydocstyle and it did pick up on the active mood so that's done -- I also picked up on a couple extra interesting checks.

```
exercises/concept/card-games/lists.py:1 at module level:
        D100: Missing docstring in public module
exercises/concept/card-games/lists.py:2 in public function `get_rounds`:
        D202: No blank lines allowed after function docstring (found 1)
```

Whatya think? I think the top level docstring is needed, as I do remember having to manually input this a couple times while doing exercises. The second part is a style decision and isn't a part of the analyzer as far as I know. Let me know and I can make these additional changes.

**Comment by BethanyG** [comment]
@Metallifax   -- Thank you so much for being a good sport around this!  As you can see, docstrings are not always well done -- even by people who claim to follow the standard. 😆   So I am happy that you've taken this on, and that you are keeping us honest.  😄 

Yes - I agree that having or not having an extra line after a function docstring in a module is more a style thing than a hard requirement. I personally default to an extra line because I feel it is more readable -- but am more than happy to change that if we decide we want to omit it. Either way works for me, we just need to make it consistent.

As for the module-level docstrings: Yes, I did leave that nag on for PyLint in the analyzer, although it can be turned off per exercise where we feel it's excessive.  So turning it off is one strategy, if we decide it's just too much to do. For most code that isn't a library or API, most developers will say that a module level docstring is excessive. Then again, most will also complain that docstrings are excessive in general. 😉 

For a module-level docstring, from a best practices for teaching standpoint, PEP 257 does indeed say:

> _All modules should normally have docstrings, and all functions and classes exported by a module should also have docstrings. Public methods (including the __init__ constructor) should also have docstrings. A package may be documented in the module docstring of the __init__.py file in the package directory._

Of course, the whole PEP is prefaced by:

>_"A universal convention supplies all of maintainability, clarity, consistency, and a foundation for good programming habits too. What it doesn't do is insist that you follow it against your will. That's Python!"
>
> —Tim Peters on comp.lang.python, 2001-06-16_

and 

> _If you violate these conventions, the worst you'll get is some dirty looks. But some software (such as the [Docutils](http://docutils.sourceforge.net/) docstring processing system [PEP 256](https://www.python.org/dev/peps/pep-0256), [PEP 258](https://www.python.org/dev/peps/pep-0258)) will be aware of the conventions, so following them will get you the best results._


Soooo. 😄   I think brief module-level docstrings for now with the exercises we have, and maybe omitting them further down the track (_along with function/module docstrings_), but leaving the nags on so that students begin to practice making them on their own.   _Sound good?_

**Comment by Metallifax** [comment]
@BethanyG Hey, this is what I signed up for, I'm kicking myself for missing it myself -- I'm just happy Isaac looked at it when he did.

As for a stopping point for docstrings, just so I'm clear, you're suggesting removing docstrings on later concept exercises for practice purposes on the part of the student -- did you have a stopping point in mind? Also, do you mean to remove all docstrings triple quotes and all, or just the summary first line sentence and module level docstrings? One benefit of the second approach is that we don't have to touch the exercises after a certain point from a maintenance perspective and still bug the students to fill it out when they get to the analyzer. Here's a list of the order of the concepts in the concept tree read row by row as reference.

```
Concept List

Guido's Gorgeous Lasagna     	Basics
Currency Exchange  		Numbers
Ghost Gobble Arcade Game  	Bools
Meltdown Mitigation  		Conditionals
Little Sister's Vocabulary	Strings
Black Jack 			Comparisons
Little Sister's Essay		String Methods
Card Games			Lists
Chaitana's Colossal Coaster	List Methods
Making the Grade		Loops
Tisbury Treasure Hunt		Tuples
Inventory Management		Dicts
Cater Waiter			Sets
Ellen's Alien Game		Classes


Work in Progress Concepts

Pretty Leaflet 			String Formatting
Log Levels			Enums
Restaurant Rozalynn		none
```

Let me know where on the list you'd like to leave docstrings out I can add that to my notes. Also, I don't mind making a Google doc to keep track of this as well, or learn ways of doing something like this in an issue as an assignment, as I'm always open for learning new things on here.

I'll get the ball rolling on the module level docstrings, looking forward to your feedback.

**Comment by BethanyG** [comment]
> _As for a stopping point for docstrings, just so I'm clear, you're suggesting removing docstrings on later concept exercises for practice purposes on the part of the student -- did you have a stopping point in mind?_

Looking at that tree in `reference/track_exercises_overview.md` (_which I am in the process of [updating](https://github.com/BethanyG/python/blob/exercise-tree-update-2-16-22/reference/track_exercises_overview.md)_), I think that `classes` (_Ellen's Alien Game_) (_which doesn't have function docstrings anyways because the exercise is to implement the class_) and `functions` (TBD) out on the other main branch make sense as stopping points. 

So that means docstrings  for `Pretty Leaflet ` (_and all the other beta status exercises except `Ellens Alien Game`_)  and no docstrings for `Log Levels` or `Restaurant Rozalynn`.	

My logic is that `classes` and `functions` are the points where a student might go from learning things like core data structures and core types to more detail and nuance of putting together longer scripts or programs or dealing with modules, etc.  Yes - they will still be doing small contained exercises on the track - but maybe also thinking about how things go together in long(er) form. But I am happy to revise my opinion if that doesn't feel good.

> _Also, do you mean to remove all docstrings triple quotes and all, or just the summary first line sentence and module level docstrings?_

I am ... a bit ambivalent?  Either way, we could nag the student.  If we omit them, we can use the existing PyLint rule and it will complain about the missing docstrings at each level.  Right now, each exercise has a custom "checker" -- so the rule can be turned off there. It's messy (_I need to refactor it to be cleaner and more config driven_) - but that's ok.  Eventually, we might be able to selectively run `pydocstyle` as a set of analyzer checks -- or fold `pydocstyle` rules into the PyLint set we have.

> _Also, I don't mind making a Google doc to keep track of this as well, or learn ways of doing something like this in an issue as an assignment, as I'm always open for learning new things on here_

_Careful what you wish for!_  😆   I've been doing some summary tracking in the `/reference` directory (also linked above), just to have a doc that shows where we think we're headed.  Right now, I am in the middle of an [update](https://github.com/BethanyG/python/blob/exercise-tree-update-2-16-22/reference/track_exercises_overview.md) to show our "new" exercises (_Black Jack and Ellens Alien Game_).  It might make sense for us to eventually add a column in that table which tracks which stubs have docstrings and which don't.   **BUT**

1.  I haven't pushed my update.  Which I need to do.
2. I wouldn't want to inflict that markdown table on anyone.  It's .... indescribable in its horror, and a beast to edit.  The whole page is, actually.  Not that I am unwilling to show you -- just that it may not be a high-leverage learning activity for you.  I'll let you decide on that.  😉 

When it comes to doing tracking issues -- I think everyone is a little different.  I tend to do things like [this issue](https://github.com/exercism/python/issues/2563) when I am working across exercises on a specific change - so that there is a general description of the change being made, as well as notes about why/where its being made, and which PRs are connected to it. But go with what communicates your intent the best, and what won't make you crazy.  And if that's individual issues as you get time to make changes -- I am happy with that. 

Just let me know -- happy to show you my markdown tricks (_and my grandma style 1999-era HTML to get them to work_) as well as any mermaid you want to play with (_mermaid.js is what the tree is written in_).  But don't feel obliged.  If you just wanted to do a tracking issue with the edits you're planning out - that would be _awesome_.  And of course, if it keeps you sane -- please do track in a google sheet.  _I have hundreds._  😉 

💙 💛   for reading to the end of this novel.  😄

File: exercises/concept/card-games/.meta/exemplar.py [hunk 1]
```diff
@@ -31,7 +37,7 @@ def list_contains_round(rounds, number):
 
 
 def card_average(hand):
-    """Calculates and returns the average card value from the list.
+    """Calculate and returns the average card value from the list.
```
Comment (BethanyG) [inline exercises/concept/card-games/.meta/exemplar.py:40]:
```suggestion
    """Calculate and return the average card value from the list.
```

**Review by BethanyG — APPROVED** [review]
Just one niggle.  Approving otherwise.  😄 ****

**Comment by Metallifax** [comment]
@BethanyG Sounds good on the exercise choices, I'll keep these selections in mind going forward and we can always just play it by ear in regards to if we want to keep docstrings in your selections when the time comes.

I was going to mention maybe using a manual graphic tool like LucidChart but it looks like you guys got it covered with mermaid.js -- the tree graphic looks great and seems way less of a headache to work on team wise (despite the humongous amount of markup).

[_Figure 1 -_](https://static.tvtropes.org/pmwiki/pub/images/steve_harvey_reaction.jpg) me scrolling through the markup view on track_exercises_overview.md

Well... I can do a quick spreadsheet for myself in the meantime to hold us over until a more permanent solution can be made in regards to the overview file. I'll plug away at the docstrings themselves when I have the time and while I can handle a little markdown myself, as you said, there seems to be a lot of moving parts there that I'll need to be in the loop for so we'll cross that bridge when it comes.

I'll go with your suggestion in regards to issue tracking and just make individual pull requests until I feel more confident I can manage things on here, I also really like the idea of an extra column on track_exercises_overview.md, so that others can be involved in keeping up on these changes.

As always, thank you for the guidance and the info, and I'll keep an eye on the overview.md update in the meantime. Cheers!

**Pull request merged**